### PR TITLE
<format>: Implement P2418R2 support for `std::generator`-like types

### LINKED ISSUE #2239: P2418R2 Add Support For `std::generator`-like Types To `std::format`
[P2418R2](https://wg21.link/P2418R2) Add Support For `std::generator`-like Types To `std::format`

This increases the value of `__cpp_lib_format`, also increased by [P2372R3](https://wg21.link/P2372R3) (see #2237).

### PR DESCRIPTION
**Pull request by barcharcraz**
Implements P2418R2 which prepares format to be able to format std::generators, by allowing "format" functions that take their parameters by non-const reference. This DR is somewhat high risk because it modifies the argument type mapping machinery.

* make(w)_format_args was modified to forward its arguments, I _think_ this is what the standard says to do but I'm not 100% sure since the standard has make_format_args directly initialize each basic_format_arg while we need to pass the arguments to all together to _Format_arg_store.
* I have modified usages of `_Storage_type` to use `remove_cvref_t<T>` instead of `T` because that's what the "other end" (the un-erase from handle) uses now. I don't think this has actual effect on the mapping, but could change the selected formatter specialization for custom types. Tim Song indicated on the reflectors that the lack of `remove_cvref_t<T>` in [format.arg]/5 is probably a defect.
* The `_Has_formatter` and the new `_Has_const_formatter` concepts have been moved up since they are now needed in the constructor for `basic_format_arg::handle`

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2418r2.html

Fixes #2239.

### CODE REVIEW

File: tests/std/tests/P0645R10_text_formatting_custom_formatting/test.cpp
Comment (StephanTLavavej) [inline tests/std/tests/P0645R10_text_formatting_custom_formatting/test.cpp:165]: Would it be worthwhile to:

* (moderate size) Have test coverage for each family of updated functions? (`format`, `format_to`, `format_to_n`, `formatted_size`, `make_format_args`)
* (large size) Have test coverage for all overloads (`char`/`wchar_t` and plain/`locale`)?

This would have detected `format_to` being missed. However, if writing such test code would be tedious/low-value, I could be persuaded that the current test coverage is sufficient (as each overload didn't need extensive fragile work, just a signature change, combined with larger changes to the common underlying machinery).

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/format
Comment (barcharcraz) [reply stl/inc/format:281]: I think this is covered by the requirement that remove_cvref_t<T> be formattable to be put in a handle (which is not what this code does, because said wording was a follow up issue (LWG3631).

If remove_cvref_t<T> still has const qualifiers then const-formattable is the same as formattable (again, that's my understanding)

File: tests/std/tests/P0645R10_text_formatting_custom_formatting/test.cpp
Comment (barcharcraz) [reply tests/std/tests/P0645R10_text_formatting_custom_formatting/test.cpp:165]: It's tedious, but not unreasonably so for just the "custom type" tests

File: stl/inc/format
Comment (barcharcraz) [reply stl/inc/format:281]: I wouldn't mind some input from Casey here

File: stl/inc/format
Comment (StephanTLavavej) [reply stl/inc/format:171]: (Or if we merge this PR first, we need to hold off on updating the macro before #2221 is merged.)

File: stl/inc/format
Comment (StephanTLavavej) [reply stl/inc/format:281]: Resolving as this was a real issue (yay! :smile_cat:) that has been fixed by a commit.

**Comment by StephanTLavavej** [comment]
@barcharcraz This is ready to have merge conflicts resolved.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! I went ahead and pushed a batch of changes to address my feedback after running all tests.

File: tests/std/tests/P0645R10_text_formatting_formatting/test.cpp
Comment (CaseyCarter) [inline tests/std/tests/P0645R10_text_formatting_formatting/test.cpp:1003]: `forward` is extraneous here - no formatter is or ever should be sensitive to value category. (Also on 1011 and 1019.)

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/format
Comment (barcharcraz) [reply stl/inc/format:2961]: I'm forwarding because libfmt forwards here, I think you're right though and they forward everywhere but then specifically undo what the forwarding would have done when `make_arg` binds them all to the fmt equivalent to `const _Storage_type<remove_cvref_t<T>>&`

>Have you submitted an LWG issue to fix the preconditions on these functions to strip cvref-quals from the Ti?

no, but Tim Song did.

Comment (barcharcraz) [reply stl/inc/format:2961]: wait is this a separate issue from LWG-3631?

File: tests/std/tests/P0645R10_text_formatting_custom_formatting/test.cpp
Comment (barcharcraz) [reply tests/std/tests/P0645R10_text_formatting_custom_formatting/test.cpp:75]: yeah, this format method is bogus, but "}" is never the spec string, it's that [pc.begin(), pc.end()] is from the start of the format spec to the end of _the entire format string_, so finding '}' is just the end of the spec

(why yes, this does allow you to "eat" replacement fields in formatters in theory, but that's the entire point, and the parser will blow up immediately with format_error (or a compile failure) if you try and do that, what it allows is stuff like `{:}aaaa}bbbbbbb` where `}aaaa` is the format specifier for the first (and only) argument. it should also allow nesting, for example
`
std::format(R"(
{:
int main() {
   std::string hello = "hello";
   fputs(stdout, std::format("{:*20}\n", hello);
   return 5;
}
}
)", some_jit_or_interpreter_context{});
`
could return something "reasonable"

Comment (barcharcraz) [reply tests/std/tests/P0645R10_text_formatting_custom_formatting/test.cpp:75]: the begin()==end() condition is "no format specifiers", in this case begin==end and they are null iterators (you get a default initialized parse context)

Comment (barcharcraz) [reply tests/std/tests/P0645R10_text_formatting_custom_formatting/test.cpp:75]: so, actually I think this is a _valid_ formatter, because it throws an error when the whole range is not parsed (since no format specifiers are allowed)

File: tests/std/tests/P0645R10_text_formatting_custom_formatting/test.cpp
Comment (barcharcraz) [reply tests/std/tests/P0645R10_text_formatting_custom_formatting/test.cpp:80]: hmm yes, I suppose, though this is testing both sorts of types (format should never be doing either copies or moves except of built-in types, so maybe we should make this type neither copyable nor moveable

File: tests/std/tests/P0645R10_text_formatting_formatting/test.cpp
Comment (barcharcraz) [reply tests/std/tests/P0645R10_text_formatting_formatting/test.cpp:1003]: the forward isn't for the formatter, but because all types must match exactly for `fmt` to be a valid parameter to `format_to_n`

File: tests/std/tests/P0645R10_text_formatting_custom_formatting/test.cpp
Comment (barcharcraz) [reply tests/std/tests/P0645R10_text_formatting_custom_formatting/test.cpp:80]: OK renamed and deleted the move constructor as well

File: stl/inc/format
Comment (barcharcraz) [reply stl/inc/format]: actually we should test this too.

File: stl/inc/format
Comment (StephanTLavavej) [inline stl/inc/format:1428]: Comment nitpick: We conventionally say something like "per the proposed resolution of LWG-nnnn" to provide a bit more context.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I pushed a conflict-free merge with `main`, in addition to a comment change, because there was the major toolset update containing Clang 13. I verified that no additional reformatting was necessary.

File: stl/inc/format
Comment (jovibor) [inline stl/inc/format:175]: Why these requires' parameters have reference type? Isn't it redundant?

> parameter-list — a comma-separated list of parameters like in a function declaration, except that default arguments are not allowed and it cannot end with an ellipsis (other than one signifying a pack expansion). **These parameters have no storage, linkage or lifetime, and are only used to assist in specifying requirements.** These parameters are in scope until the closing } of the requirement-seq

https://en.cppreference.com/w/cpp/language/constraints

File: tests/std/tests/P0645R10_text_formatting_custom_formatting/test.cpp
Comment (CaseyCarter) [reply tests/std/tests/P0645R10_text_formatting_custom_formatting/test.cpp]: Nice.

File: tests/std/tests/P0645R10_text_formatting_formatting/test.cpp
Comment (CaseyCarter) [reply tests/std/tests/P0645R10_text_formatting_formatting/test.cpp:1028]: Ah - the value category is encoded in `fmt`? We'll worry about this in a year or so when we're making a throughput pass.

File: stl/inc/format
Comment (CaseyCarter) [reply stl/inc/format]: Yes it is - [format.arg.store]/2. We can do this as a followup - it doesn't need to block merging.

File: stl/inc/format
Comment (CaseyCarter) [reply stl/inc/format:175]: They inhibit array-to-pointer and function-to-pointer decay, and they ensure that e.g. `decltype((_Val))` is an lvalue reference type whether `_Ty` is a reference or not. In this particular use, we only care about the decay.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for implementing the last C++20 DR! :tada: :heart_eyes_cat: :white_check_mark:

**Pull request merged**
